Game/Plugins/MetricsServiceProvider/Source/MetricsServiceProvider/Private/MetricsServiceModule.cpp

```diff
@@ -0,0 +1,640 @@
+	{
+		// custom handling for EditorTelemetry events because those are passed from the epic side
+		// and have no notion of the MetricsBlueprintLibrary event classes
+
+		// our camelcasing is naive, so it'll camelcase abbreviation only fields (DDC -> dDC)
+		// this is cribbed from MetricsBlueprintLibrary::EnumToString
+		FString EventClass = StaticEnum<EMetricsClass>()->GetValueAsString(EMetricsClass::MetricsClass_EditorTelemetry);
+		EventClass.RemoveFromStart("EMetricsClass::MetricsClass_");
+		ConvertToCamelCase(EventClass);
+
+		// the current format from epic -> io for editor telemetry is MetricsEditorTelemetry_EventName
+		FString ParsedEventName = FString::Printf(TEXT("%s"), *EventName);
+		ParsedEventName.RemoveFromStart(GEditorTelemetryPrefix);
+		ConvertToCamelCase(ParsedEventName);
+
+		// use the windows login name; this should be internal only reporting
+		const FString ProfileID = UKismetSystemLibrary::GetPlatformUserName();
+
+		TelemetryClassEvent(EventClass, ParsedEventName, ProfileID, Attributes);
+	}
+	else
+	{
+		// Pass through to our main telemetry reporting endpoint
+		TelemetryClassEvent(TEXT(""), EventName, "", Attributes);
+	}
+}
+
+// Adds accountid/profileid/client and server session id details as appropriate.
+void FAnalyticsProviderMetrics::AugmentPayloadWithSessionDetails(MetricsPayload& PayloadData, const FString& EventProfileID)
+{
+	auto AugmentClientDetailsFromProfile = [&PayloadData, this](const FString& ProfileID) {
+		// If we have profile info, we will add it.  We cache this information when we select a profile.
+		if (const FProfileDetails* ProfileDetails = CachedProfileDetails.Find(ProfileID))
+		{
+			PayloadData.ProfileID = ProfileID;
+			PayloadData.AccountID = ProfileDetails->AccountID;
+			PayloadData.ClientSessionID = ProfileDetails->ClientSessionID;
+		}
+	};
+
+	if (GWorld)
+	{
+		switch (GWorld->GetNetMode())
+		{
+			// Our two server cases.
+			case NM_DedicatedServer:	// fallthrough
+			case NM_ListenServer:
+				// Store the profile of the character who generated this event.  Possibly blank
+				PayloadData.ServerSessionID = SessionId;
+
+				AugmentClientDetailsFromProfile(EventProfileID);
+				break;
+
+			// While we are in the main menu, before connecting as a client.  We have an account ID, but no profile id for part of this flow.
+			// So if we can't find the cached profile details we skip the profile ID data and just use the account ID
+			case NM_Standalone:
+				if (const FProfileDetails* ProfileDetails = CachedProfileDetails.Find(EventProfileID))
+				{
+					PayloadData.ProfileID = EventProfileID;
+					PayloadData.AccountID = ProfileDetails->AccountID;
+					PayloadData.ClientSessionID = ProfileDetails->ClientSessionID;
+				}
+				else
+				{
+					PayloadData.ClientSessionID = SessionId;
+					PayloadData.AccountID = EventProfileID;
+				}
+				break;
+
+			// Clients connected to servers.
+			case NM_Client:
+				PayloadData.ClientSessionID = SessionId;
```

Comment:
It should be generated in `StartSession` method, and in `GDKTestGymsGameInstance`'s `Init` method we will call this function to generate an unique code and then we can override here.

Reply:
Cool. I guess I ask because PayloadData.ClientSessionID might be set inside AugmentClientDetailsFromProfile. So PayloadData.ClientSessionID might end up being ProfileDetails->ClientSessionID instead of SessionId. I am not sure what the intended behavior is here. What should take precedence?